sonar-kotlin-checks/src/main/java/org/sonarsource/kotlin/checks/SimplifyFilteringBeforeTerminalOperationCheck.kt

```diff
@@ -0,0 +1,64 @@
+/*
+ * SonarSource Kotlin
+ * Copyright (C) 2018-2023 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonarsource.kotlin.checks
+
+import org.jetbrains.kotlin.psi.KtCallExpression
+import org.jetbrains.kotlin.psi.KtDotQualifiedExpression
+import org.jetbrains.kotlin.resolve.calls.callUtil.getCall
+import org.jetbrains.kotlin.resolve.calls.model.ResolvedCall
+import org.jetbrains.kotlin.utils.IDEAPluginsCompatibilityAPI
+import org.sonar.check.Rule
+import org.sonarsource.kotlin.api.checks.CallAbstractCheck
+import org.sonarsource.kotlin.api.checks.FunMatcher
+import org.sonarsource.kotlin.api.frontend.KotlinFileContext
+
+private const val KOTLIN_COLLECTIONS_QUALIFIER = "kotlin.collections"
+private val FILTER_MATCHER = FunMatcher(qualifier = KOTLIN_COLLECTIONS_QUALIFIER, name = "filter") { withArguments("kotlin.Function1") }
+
+@Rule(key = "S6527")
+class SimplifyFilteringBeforeTerminalOperationCheck : CallAbstractCheck() {
+    override val functionsToVisit = listOf(
+        FunMatcher(qualifier = KOTLIN_COLLECTIONS_QUALIFIER) {
+            withNames("any", "none", "count", "last", "lastOrNull", "first", "firstOrNull", "single", "singleOrNull")
+            withNoArguments()
+        }
+    )
+
+    @OptIn(IDEAPluginsCompatibilityAPI::class)
+    override fun visitFunctionCall(callExpression: KtCallExpression, resolvedCall: ResolvedCall<*>, kotlinFileContext: KotlinFileContext) {
+        val chainedCallBeforeTerminalOperationCall = callExpression.parent
+            .let { it as? KtDotQualifiedExpression }
+            ?.receiverExpression
+            .let { it?.getCall(kotlinFileContext.bindingContext) }
+
+        if (chainedCallBeforeTerminalOperationCall != null
```

Comment:
You can avoid this null-check by returning early while retrieving the `chainedCallBeforeTerminalOperationCall`. Just add a `?: return` after line 49 and you will be guaranteed that you cannot arrive here with `chainedCallBeforeTerminalOperationCall` being null.

Then, if you like functional style, you don't even need that. Using the `.takeIf` function, you can chain everything and don't need an explicit `if`, without losing readability. So something like:
```
...
?.receiverExpression
?.getCall(...)
?.takeIf { FILTER_MATCHER.matches(...) }
?.let { chainedCallBeforeTerminalOperationCall ->
  // the reporting logic
}
```
It's also fine to leave a separate `if` conditional to check if it matches, though.